### PR TITLE
*: support adding role in auth v3

### PR DESCRIPTION
**Pull request by mitake**

### CODE REVIEW

File: auth/authpb/auth.proto [hunk 1]
```diff
@@ -15,3 +15,10 @@ message User {
   bytes password = 2;
   int64 tombstone = 3;
 }
+
+// Role is a single entry in the bucket authRoles
+message Role {
+  int64 ID = 1;
```
Comment (xiang90) [inline auth/authpb/auth.proto]: why do we need ID? we want to allow two roles have same name? I feel name should just be the primary key.

Comment (mitake) [inline auth/authpb/auth.proto]: For example, users can 1. create role A and grant A to multiple users 2. delete A 3. create a new role A. If we don't use a role ID, we need to update every (possibly many) user that is granted role A during step 2. Otherwise, users that granted A during step 1 will get the new role A created during step 3 illegally.

So I decided to use current scheme of allocating ID during creation and marking with tombstone during deletion (described in the doc). How do you think?

Comment (xiang90) [inline auth/authpb/auth.proto]: > For example, users can 1. create role A and grant A to multiple users 2. delete A 3. create a new role A. If we don't use a role ID, we need to update every (possibly many) user that is granted role A during step 2. Otherwise, users that granted A during step 1 will get the new role A created during step 3 illegally.

If we treat A as the identification of a role, then it is legal. Once we remove role A, then user with role A just treat role A as an empty role. When we add role A back, then role A becomes non-empty again.

Basically we allow user to have empty roles attached.

Comment (xiang90) [inline auth/authpb/auth.proto]: And remove a role will not necessarily remove the role from users.

Comment (mitake) [inline auth/authpb/auth.proto]: > Basically we allow user to have empty roles attached.

I didn't notice this behavior of v2.

However, the model would introduce unintentional permission configuration. At least we should provide an option of `role delete` that deletes a role from entire users, I think.

Comment (xiang90) [inline auth/authpb/auth.proto]: that is user revoke role. We should be able to revoke
 Roles from users

Comment (mitake) [inline auth/authpb/auth.proto]: Yes, but there's no way to revoke deleted roles from entire users that are granted the roles. I think providing the way would be safer for auth management.

**Comment by mitake** [comment]
@xiang90 updated for using Name as a key of Role, PTAL.

File: auth/store.go [hunk 2]
```diff
@@ -25,14 +25,19 @@ import (
...
+
 	authBucketName      = []byte("auth")
 	authUsersBucketName = []byte("authUsers")
+	authRolesBucketName = []byte("authRoles")
 
 	plog = capnslog.NewPackageLogger("github.com/coreos/etcd", "auth")
 
 	ErrUserAlreadyExist = errors.New("auth: user already exists")
 	ErrUserNotFound     = errors.New("auth: user not found")
+	ErrBrokenAuthStore  = errors.New("auth: state of auth store is broken")
```
Comment (xiang90) [inline auth/store.go]: remove this error since no one is using it?

File: etcdctl/ctlv3/command/role_command.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+		Use:   "role <subcommand>",
+		Short: "role related command",
+	}
+
+	ac.AddCommand(NewRoleAddCommand())
+
+	return ac
+}
+
+func NewRoleAddCommand() *cobra.Command {
```
Comment (xiang90) [inline etcdctl/ctlv3/command/role_command.go]: comment for the public func?

File: auth/authpb/auth.proto (hunk 1 above)
Comment (xiang90) [inline auth/authpb/auth.proto]: We could, probably by adding an expensive version of role delete which also traverse users to remove attached roles. also we can build internal index to speed this up if we want.

**Comment by xiang90** [comment]
LGTM. But do you agree with https://github.com/coreos/etcd/pull/4926#discussion_r58449480?

File: etcdctl/ctlv3/command/role_command.go (hunk 3 above)
Comment (mitake) [inline etcdctl/ctlv3/command/role_command.go]: On the second thought, the function doesn't need to be public. I'll make it private.

**Comment by mitake** [comment]
@xiang90 updated for the comment, PTAL. I agree with the comment.

**Comment by xiang90** [comment]
lgtm!
